### PR TITLE
VAULT-17079: Adding Hash Function and HeaderAdjuster to EntryFormatter

### PR DESCRIPTION
**Pull request by marcboudreau**
This PR add an additional field to the **EntryFormatter** structure, which acts as the formatter Node in [go-eventlogger](https://github.com/hashicorp/go-eventlogger) Pipelines used for persisting audit records. This new field is an instance of the HeaderAdjuster interface.

The **HeaderAdjuster** interface requires a single function, **ApplyConfig**. The function signature matches the signature of the **ApplyConfig** method of the **AuditedHeadersConfig** structure.  

This change is laying the ground work to be able to flow **LogInput** records as events through the [go-eventlogger](https://github.com/hashicorp/go-eventlogger) framework. The **ApplyConfig** method of the **AuditedHeadersConfig** structure is currently called repeatedly for each element in the **AuditBroker** structure's **backends** field in the **LogRequest** and **LogResponse** methods of the **AuditBroker** structure. This function call examines the request headers and cross checks them with a configured list of headers that should be included in the audit record. The header may also need to be hashed according to the **hmac** parameter in the configuration. Each audit backend has its own hashing function that uses the backend's unique salt.

As we move towards using the [go-eventlogger](https://github.com/hashicorp/go-eventlogger), a single event will be sent to a Broker to deliver it through every Pipeline registered for the event's type. This architecture will make it impossible to pre-process the audit records prior to sending them to the broker. Instead, the header manipulation will need to be done within each pipeline. The anticipated design will have that manipulation done in the pipeline's formatter node, which is why a pointer to a structure implementing the **HeaderAdjuster** interface is being added here.

Once the above was achieved, quite a few functions and methods involved had both a parameter of type **audit.Salter** and a parameter of type `func(context.Context, string) (string, error)` which was used to take a backend's GetHash method. Both of these parameters are related and the implementation of the **GetHash** function is so trivial and it relies on the **audit.Salter** parameter. This meant that wherever a parameter of type `func(context.Context, string) (string, error)` was needed, it could be replaced with a parameter of type **audit.Salter**. This would eradicate the need for both parameters, which eliminate the need to add yet more function/method parameters.

### CODE REVIEW

File: audit/entry_formatter.go [hunk 1]
```diff
@@ -48,19 +48,20 @@ func NewEntryFormatter(config FormatterConfig, salter Salter, opt ...Option) (*E
...
+		salter:        salter,
+		config:        config,
+		headersConfig: headersConfig,
+		prefix:        opts.withPrefix,
 	}, nil
 }
 
 // Reopen is a no-op for the formatter node.
-func (_ *EntryFormatter) Reopen() error {
+func (*EntryFormatter) Reopen() error {
```
Comment (marcboudreau) [inline audit/entry_formatter.go:59]: Unrelated cleanup of a go-staticcheck warning.

File: audit/entry_formatter.go [hunk 2]
```diff
@@ -48,19 +48,20 @@ func NewEntryFormatter(config FormatterConfig, salter Salter, opt ...Option) (*E
...
 	return nil
 }
 
 // Type describes the type of this node (formatter).
-func (_ *EntryFormatter) Type() eventlogger.NodeType {
+func (*EntryFormatter) Type() eventlogger.NodeType {
```
Comment (marcboudreau) [inline audit/entry_formatter.go:64]: Unrelated cleanup of a go-staticcheck warning.

File: audit/entry_formatter_test.go [hunk 3]
```diff
@@ -366,13 +372,13 @@ func BenchmarkAuditFileSink_Process(b *testing.B) {
 		},
 	}
 
-	ctx := namespace.RootContext(nil)
+	ctx := namespace.RootContext(context.Background())
```
Comment (marcboudreau) [inline audit/entry_formatter_test.go:375]: Unrelated cleanup of go-staticcheck warning.

File: vault/audited_headers_test.go [hunk 4]
```diff
@@ -181,20 +188,40 @@ func TestAuditedHeadersConfig_ApplyConfig(t *testing.T) {
...
+		"x-vault-header": {"hmac-sha256:", "hmac-sha256:"},
+	}
+
+	if len(expected) != len(result) {
+		t.Fatalf("Expected headers count did not match actual count: Expected count %d\n Got %d\n", len(expected), len(result))
 	}
 
-	if !reflect.DeepEqual(result, expected) {
-		t.Fatalf("Expected headers did not match actual: Expected %#v\n Got %#v\n", expected, result)
+	for resultKey, resultValues := range result {
```
Comment (peteski22) [inline vault/audited_headers_test.go:207]: ❓ would we be able to pull out some decent test cases to exercise the code in lots of ways? The pre-existing test just seems to be a happy path test and it would be good to be confident that we've tried to break things and seen it happen in ways we expect.

**Review by peteski22 — COMMENTED** [review]
Thanks Marc! I added some comments, I hope some are helpful, and some aren't too annoying. 😅

File: audit/entry_formatter.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ var (
 
 // NewEntryFormatter should be used to create an EntryFormatter.
 // Accepted options: WithPrefix.
-func NewEntryFormatter(config FormatterConfig, salter Salter, opt ...Option) (*EntryFormatter, error) {
+func NewEntryFormatter(config FormatterConfig, salter Salter, headersConfig HeaderAdjuster, opt ...Option) (*EntryFormatter, error) {
```
[the remark this answers is not in the document]
Comment (marcboudreau) [reply audit/entry_formatter.go:33]: I think that the **EntryFormatter** can operate even if the headersConfig is `nil`. It would just mean that the step of examining the request headers would be skipped. This is what happens when the LogTestMessage is called.

File: audit/entry_formatter_test.go [hunk 6]
```diff
@@ -144,13 +144,19 @@ func TestNewEntryFormatter(t *testing.T) {
 	}
 }
 
+type TestingHeadersAdjuster struct{}
```
[the remark this answers is not in the document]
Comment (marcboudreau) [reply audit/entry_formatter_test.go:147]: Actually, rather than create a no-op implementation of **HeadersAdjuster**, the **headersConfig** parameter can be set to `nil` in the tests. So, I'm going to remove this struct for now and later on when the **headersConfig** field is actually used and tests that rely on it are introduced, I'll bring it back.

File: audit/hashstructure_test.go [hunk 7]
```diff
@@ -98,20 +98,28 @@ func TestCopy_response(t *testing.T) {
 	}
 }
 
-func TestHashString(t *testing.T) {
+type TestSalter struct{}
```
[the remark this answers is not in the document]
Comment (marcboudreau) [reply audit/hashstructure_test.go:101]: go doc comments: 👍 
question ❓ : I don't think that is necessary, since we are likely to use the **TestSalter** type somewhere in the same set of tests that a **Salter** interface is expected. If **TestSalter** doesn't satisfy the interface, then it will generate a compiler error. Using the `var _ Salter = (*TestSalter)(nil)` expression would be done if the TestSalter struct was defined in a module but not used as a **Salter** interface in that module. In that case, there wouldn't be any compiler errors when building the module where **TestSalter** is defined. However, when some other code imports the **TestSalter**'s module and tries to use a **TestSalter** instance where a **Salter** interface is required, that would generate a compiler error.

File: audit/types.go [hunk 8]
```diff
@@ -85,11 +85,19 @@ type Writer interface {
 	WriteResponse(io.Writer, *ResponseEntry) error
 }
 
+type HeaderAdjuster interface {
+	// ApplyConfig returns a map of header values that consists of the
+	// intersection of the provided set of header values with a configured
+	// set of headers and will hash headers that have been configured as such.
+	ApplyConfig(context.Context, map[string][]string, Salter) (map[string][]string, error)
```
[the remark this answers is not in the document]
Comment (marcboudreau) [reply audit/types.go:92]: I agree with you that the name poorly conveys the behaviour of this method, but I'd like to put off changing that name to a future PR in order to keep the scope under control.

File: vault/audited_headers_test.go (hunk 4 above)
Comment (marcboudreau) [reply vault/audited_headers_test.go:207]: I can add test functions for the case where there's no request headers and where an error is returned by the HashString function.

**Review by peteski22 — APPROVED** [review]
